tests/unit/s2n_early_data_test.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+#include "tls/s2n_early_data.h"
+
+typedef struct {
+    const s2n_early_data_state *states;
+    size_t len;
+} s2n_early_state_sequence;
+
+/* We want to verify that applying s2n_connection_set_early_data_state to the current state
+ * can only produce valid state sequences.
+ *
+ * We check every possible next state and verify that the only transitions that
+ * s2n_connection_set_early_data_state allows are those in the valid state sequences. Then, for
+ * every valid transition, we call this method again recursively. The recursion ends when we either
+ * reach the end of a valid state sequence or encounter an invalid state sequence.
+ */
+static S2N_RESULT s2n_test_all_early_data_sequences(struct s2n_connection *conn, size_t index,
```

Comment:
Isn't "index" one of those reserved words that will cause problems for older systems?

Reply:
I added an issue to automatically detect this mistake: https://github.com/awslabs/s2n/issues/2591